src/cache_invalidate.c

```diff
@@ -67,6 +67,7 @@ static void
 cache_invalidate_callback(Datum arg, Oid relid)
 {
 	Catalog *catalog;
+	static bool in_recursion = false;
```

Comment:
Is there a risk that multiple threads might be trying to invalidate items at the same time?

Reply:
PostgreSQL is single-threaded: there is just one process that executes. This pattern is used quite often inside PostgreSQL itself, for example in `src/backend/utils/cache/inval.c`.